sled-agent/src/hardware/illumos/nvme.rs

```diff
@@ -0,0 +1,118 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! ioctls for querying NVMe devices
+//!
+//! For more context on the structures referenced in this file, refer to:
+//! <https://github.com/illumos/illumos-gate/blob/98f586d71279849001034981c5906e9e3901d56f/usr/src/uts/common/sys/nvme.h>
+
+use std::fs::OpenOptions;
+use std::io;
+use std::os::unix::io::AsRawFd;
+use std::path::PathBuf;
+
+/// Minimal subset of the NVMe ID Controller data structure used to identify a WDC device.
+#[derive(Debug, Clone)]
+pub struct ControllerId {
+    pub vendor_id: u16,
+    pub subsystem_vendor_id: u16,
+    pub serial: String,
+    pub model: String,
+    pub firmware_rev: String,
+    pub oui: [u8; 3],
+    pub unique_controller_id: u16,
+    pub nvme_version: String,
+    pub fru_guid: String,
+}
+
+pub(crate) const NVME_ID_CTRL_BUFSIZE: usize = 4096;
+
+impl From<&[u8]> for ControllerId {
+    fn from(buf: &[u8]) -> ControllerId {
+        assert!(buf.len() >= NVME_ID_CTRL_BUFSIZE);
+        let vendor_id = u16::from_le_bytes(buf[..2].try_into().unwrap());
```

Comment:
Should we use `from_ne_bytes` throughout this function? I'm assuming `buf` contains the equivalent of `memcpy`'ing the `nvme_identify_ctrl_t` C structure, which would give us native endianness on all the fields.

Reply:
Following up on that - if you wanted to run the `nvme_identify_ctrl_t` through bindgen to get a Rust-equivalent definition, we could use [zerocopy](https://docs.rs/zerocopy/latest/zerocopy/index.html) to copy from `buf` into that struct, and then pull all the fields out directly instead of having to hardcode offsets. (We use `zerocopy` in hubris quite a lot, although more for IPC than FFI.)

Reply:
Updated to `from_ne_bytes`, agreed that makes more sense than assuming the host is little endian.

I went with the non-bindgen approach just to keep the code surface area relatively small, but I'd be open to a bindgen-produced API if we really wanted a more involved interaction with the NVME interface.